### PR TITLE
optimize reflection padding performance on CPU

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102949
* #102597
* #102518
* #102255
* __->__ #102254

This patch improves reflection padding performance on CPU.

Original kernel has nested paralleled loops, e.g. first on dim of **batch** and then on dim of **channels**, this is not optimal practice when N * C is small. This patch did dimension collapse on NC and adjacent spatial dims to maximize the parallelism scope.

The following benchmark result gathered on Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, with 20 cores per socket.

### single core inference
```
(before)
ReflectionPad2d((2, 2, 2, 2)) size:  torch.Size([1, 3, 224, 224]) , NCHW: 0.281 ms;  
ReflectionPad2d((2, 2, 2, 2)) size:  torch.Size([128, 64, 56, 56]) , NCHW: 55.675 ms; 

(after)
ReflectionPad2d((2, 2, 2, 2)) size:  torch.Size([1, 3, 224, 224]) , NCHW: 0.049 ms;  
ReflectionPad2d((2, 2, 2, 2)) size:  torch.Size([128, 64, 56, 56]) , NCHW: 17.252 ms;  
```

### single socket inference
```
(before)
ReflectionPad2d((2, 2, 2, 2)) size:  torch.Size([1, 3, 224, 224]) , NCHW: 0.118 ms;  
ReflectionPad2d((2, 2, 2, 2)) size:  torch.Size([128, 64, 56, 56]) , NCHW: 4.023 ms; 

(after)
ReflectionPad2d((2, 2, 2, 2)) size:  torch.Size([1, 3, 224, 224]) , NCHW: 0.010 ms;  
ReflectionPad2d((2, 2, 2, 2)) size:  torch.Size([128, 64, 56, 56]) , NCHW: 3.149 ms; 
```

cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/ReflectionPad.cpp [hunk 1]
```diff
@@ -922,154 +372,35 @@ Tensor reflection_pad2d_backward_cpu(
     const Tensor& grad_output,
     const Tensor& input,
     IntArrayRef padding) {
-  auto grad_input = at::zeros_like(input, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  auto grad_input = at::empty({0}, input.options());
```
Comment (cpuhrsch) [inline aten/src/ATen/native/ReflectionPad.cpp:375]: I see you're resizing the empty grad input. Maybe you mean to use empty_like?

Again I'm not sure whether this is a good change for autograd. cc @albanD

**Comment by cpuhrsch** [comment]
Overall I think this is fine. Just flagging the switch from zeros to empty. Waiting on resolution from @albanD .

File: aten/src/ATen/native/ReflectionPad.cpp (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/native/ReflectionPad.cpp]: updated!

**Comment by mingfeima** [comment]
@cpuhrsch this patch has been updated, could you please help review again ?

**Review by cpuhrsch — APPROVED** [review]

**Pull request closed without merging**